src/aspire/abinitio/commonline_cn.py

```diff
@@ -0,0 +1,376 @@
+                    corrs = np.mean(corrs, axis=1)
+                    corrs = np.reshape(
+                        corrs,
+                        (
+                            self.n_points_sphere,
+                            self.n_points_sphere,
+                            n_theta_ijs // self.order,
+                        ),
+                    )
+
+                    # Self common-lines are invariant to n_theta_ijs (i.e., in-plane rotation angles) so max them out.
+                    opt_theta_ij_ind_per_sphere_points = np.argmax(corrs, axis=-1)
+                    corrs = np.max(corrs, axis=-1)
+
+                    # Maximum likelihood while taking into consideration both cls and scls.
+                    corrs = corrs * np.outer(scores_self_corrs[i], scores_self_corrs[j])
+
+                    # Extract the optimal candidates.
+                    opt_sphere_i, opt_sphere_j = np.unravel_index(
+                        np.argmax(corrs), corrs.shape
+                    )
+                    opt_theta_ij = opt_theta_ij_ind_per_sphere_points[
+                        opt_sphere_i, opt_sphere_j
+                    ]
+
+                    opt_Ri_tilde = Ris_tilde[opt_sphere_i]
+                    opt_Rj_tilde = Ris_tilde[opt_sphere_j]
+                    opt_R_theta_ij = R_theta_ijs[opt_theta_ij]
+
+                    # Compute the estimate of vi*vi.T as given by j.
+                    vii_j = np.mean(opt_Ri_tilde.T @ rots_symm @ opt_Ri_tilde, axis=0)
+                    mean_est[i].push(vii_j)
+
+                    # Compute the estimate of vj*vj.T as given by i.
+                    vjj_i = np.mean(opt_Rj_tilde.T @ rots_symm @ opt_Rj_tilde, axis=0)
+                    mean_est[j].push(vjj_i)
+
+                    # Compute the estimate of vi*vj.T.
+                    vijs[c] = np.mean(
+                        opt_Ri_tilde.T @ rots_symm @ opt_R_theta_ij @ opt_Rj_tilde,
+                        axis=0,
+                    )
+
+                    c += 1
+                    pbar.update()
+
+                # There are conflicting methods betweeen the paper and the Matlab code for
+                # finding the optimal estimate. The paper suggests using SVD to find the
+                # estimate for vii which is closest to rank 1. The Matlab code takes the
+                # median over the stack of all estimates for vii. Here we have implemented
+                # a method which J-synchronizes the estimates prior to taking the mean.
+                # See issue #869 for more details.
+                viis[i] = mean_est[i].synchronized_mean()
+
+        # As we are using a mean to get the estimates, viis, the estimate will not be rank-1
+        # So we use SVD to find a close rank-1 approximation.
+        U, S, V = np.linalg.svd(viis)
+        S_rank1 = np.zeros((n_img, 3, 3), dtype=self.dtype)
+        S_rank1[:, 0, 0] = S[:, 0]
+        viis_rank1 = U @ S_rank1 @ V
```

Comment:
What about `S[:, 0][..., np.newaxis, np.newaxis] * (U[:, :, :1] @ V[:, :1, :])`? Maybe less clear, actually…

Reply:
I like this, but I agree that it's a little more clear how it is. Going to leave this for now.

Reply:
Sounds good. Made the method `best_rank1_approximation()` that consumes a 2D array or 3D array that's a stack and returns the rank-1 approximation. Placed it in `utils.matrix` and added an associated unit test.